### PR TITLE
cudaPackages: 11.5 -> 11.6 and gcc11

### PR DESCRIPTION
**Pull request by SomeoneSerge**
###### Description of changes

Cuda has strict requirements for the compiler and libc versions, in particular `11.5` is incompatible with glibc `2.4`, but `11.6` introduces a fix for that.

This PR updates the default cudaPackages of nixpkgs from `11.5` to `11.6`, resets preferred gcc versions for cuda `11.4` and  `11.5` back to `gcc10`, and keeps `gcc11` for cuda `11.6`.

At first glance, this seems to fix the cuda-related breakages introduced in https://github.com/NixOS/nixpkgs/pull/165732.
Hopefully this would allow us to revert the `cudaPackages_11_3` fallbacks introduced in https://github.com/NixOS/nixpkgs/pull/169279 as a fix

Waiting for the builds to finish: https://hercules-ci.com/github/SomeoneSerge/nixpkgs-unfree/jobs/449 (also a [baseline](https://hercules-ci.com/github/SomeoneSerge/nixpkgs-unfree/jobs/451))

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/cudatoolkit/versions.toml [hunk 1]
```diff
@@ -44,13 +44,13 @@ gcc = "gcc9"
...
 sha256 = "sha256-u9h8oOkT+DdFSnljZ0c1E83e9VUILk2G7Zo4ZZzIHwo="
-gcc = "gcc11"
+gcc = "gcc10"
 
 ["11.5"]
 version = "11.5.0"
 url = "https://developer.download.nvidia.com/compute/cuda/11.5.0/local_installers/cuda_11.5.0_495.29.05_linux.run"
 sha256 = "sha256-rgoWk9lJfPPYHmlIlD43lGNpANtxyY1Y7v2sr38aHkw="
-gcc = "gcc11"
+gcc = "gcc10" # cuda 11.5 has problems with glibc 2.4 -> keeping gcc10
```
Comment (samuela) [inline pkgs/development/compilers/cudatoolkit/versions.toml:53]: Would you mind also adding a link to that NVIDIA forums discussion?

**Review by samuela — COMMENTED** [review]
Love to see it!

File: pkgs/development/compilers/cudatoolkit/versions.toml (hunk 1 above)
Comment (SomeoneSerge) [reply pkgs/development/compilers/cudatoolkit/versions.toml]: There!

**Comment by samuela** [comment]
Ok I think if the builds check out then this is good to go!

**Comment by SomeoneSerge** [comment]
Ok, so this at least unbreaks `blender` and `librealsense`, and I've built `pytorch` and `cupy` before the `nccl` update - this should be safe, I guess

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 169484` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>cudaPackages.nvidia_driver</li>
    <li>truecrack-cuda</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>colmapWithCuda</li>
    <li>python39Packages.tensorflowWithCuda</li>
  </ul>
</details>
<details>
  <summary>62 packages built:</summary>
  <ul>
    <li>cudaPackages.cuda-samples</li>
    <li>cudaPackages.cuda_cccl</li>
    <li>cudaPackages.cuda_cudart</li>
    <li>cudaPackages.cuda_cuobjdump</li>
    <li>cudaPackages.cuda_cupti</li>
    <li>cudaPackages.cuda_cuxxfilt</li>
    <li>cudaPackages.cuda_demo_suite</li>
    <li>cudaPackages.cuda_documentation</li>
    <li>cudaPackages.cuda_gdb</li>
    <li>cudaPackages.cuda_memcheck</li>
    <li>cudaPackages.cuda_nsight</li>
    <li>cudaPackages.cuda_nvcc</li>
    <li>cudaPackages.cuda_nvdisasm</li>
    <li>cudaPackages.cuda_nvml_dev</li>
    <li>cudaPackages.cuda_nvprof</li>
    <li>cudaPackages.cuda_nvprune</li>
    <li>cudaPackages.cuda_nvrtc</li>
    <li>cudaPackages.cuda_nvtx</li>
    <li>cudaPackages.cuda_nvvp</li>
    <li>cudaPackages.cuda_sanitizer_api</li>
    <li>cudatoolkit (cudaPackages.cudatoolkit ,cudatoolkit_11)</li>
    <li>cudaPackages.cudnn (cudaPackages.cudnn_8_3_2)</li>
    <li>cudaPackages.cutensor</li>
    <li>cudaPackages.fabricmanager</li>
    <li>cudaPackages.libcublas</li>
    <li>cudaPackages.libcufft</li>
    <li>cudaPackages.libcufile</li>
    <li>cudaPackages.libcurand</li>
    <li>cudaPackages.libcusolver</li>
    <li>cudaPackages.libcusparse</li>
    <li>cudaPackages.libnpp</li>
    <li>cudaPackages.libnvidia_nscq</li>
    <li>cudaPackages.libnvjpeg</li>
    <li>cudaPackages.nccl</li>
    <li>cudaPackages.nsight_compute</li>
    <li>cudaPackages.nsight_systems</li>
    <li>cudaPackages.nvidia_fs</li>
    <li>ethminer (ethminer-cuda)</li>
    <li>forge</li>
    <li>gpu-burn</li>
    <li>gromacsCudaMpi</li>
    <li>gwe</li>
    <li>katagoWithCuda</li>
    <li>librealsenseWithCuda</li>
    <li>magma</li>
    <li>nvtop</li>
    <li>python310Packages.TheanoWithCuda</li>
    <li>python310Packages.cupy</li>
    <li>python310Packages.jaxlibWithCuda</li>
    <li>python310Packages.pycuda</li>
    <li>python310Packages.pynvml</li>
    <li>python310Packages.pyrealsense2WithCuda</li>
    <li>python310Packages.pytorchWithCuda</li>
    <li>python39Packages.TheanoWithCuda</li>
    <li>python39Packages.cupy</li>
    <li>python39Packages.jaxlibWithCuda</li>
    <li>python39Packages.numbaWithCuda</li>
    <li>python39Packages.pycuda</li>
    <li>python39Packages.pynvml</li>
    <li>python39Packages.pyrealsense2WithCuda</li>
    <li>python39Packages.pytorchWithCuda</li>
    <li>xpraWithNvenc</li>
  </ul>
</details>

**Comment by SomeoneSerge** [comment]
Colmap has been failing since `gcc10 -> gcc11`.
Tensorflow [has been](https://hercules-ci.com/accounts/github/SomeoneSerge/derivations/%2Fnix%2Fstore%2Fl68cll95apls11wg91bnxldb69dv38p2-tensorflow-gpu-2.8.0.drv/log?via-job=2f1c5401-ff19-437f-9f56-c5c47dba1e10) failing with this `std::__throw_bad_array_new_length()@GLIBCXX_3.4.29` too

**Comment by samuela** [comment]
Just finished running and found the same... merging now!

**Comment by dmizelle** [comment]
I ran into this today when trying to build magma! Just wanted to say thanks.

**Pull request merged**